### DISCUSSION TITLE
Retrieve temporary token from /auth/login/:provider

### DISCUSSION THREAD

**Discussion by FlorianMeinschad** [directus/directus#17010] [Archive]
I am using Keycloak as IDP for Directus and the frontend login is working fine.
I also would like to retrieve a temporary access token from the Rest API endpoint /auth/login/:provider as mentioned in the docs (https://docs.directus.io/reference/authentication.html#login).

But I get the error -> { "message": "Route /auth/login/keycloak doesn't exist.", "extensions": { "code": "ROUTE_NOT_FOUND" }}.
Querying the auth providers lists keycloak as auth provider -> {"data: [{ "name": "keycloak", "driver": "openid" }], "disableDefault": false }

How is this endpoint intended to be used?
How can I get a temporary access token for an IDP connected Directus user?

Thanks in advance!

**Comment by aidenfoxx** [comment]
The documentation you're looking at is for triggering an authentication process. With SSO flows, the endpoint behaves a little differently, and is a GET endpoint which redirects the user to the authentication page of the SSO provider, where the user can log in, and return with an `access_token`. Importantly though, this isn't a Keycloak token, but a Directus token.

**Reply by aidenfoxx** [reply]
To get a fresh access token, you should use the "refresh" endpoint, which you can send a refresh token to get back a fresh access token. https://docs.directus.io/reference/authentication.html#refresh

**Reply by ymys** [reply]
Hi, I'm new to directus. I'm already set up directus with keycloak and works fine if I access it at web directus url. 

But, now, I need API accees to integrate with my mobile app and must use Keycloak as auth. So, in login page, I set API to /auth/login/keycloak. But the problem sih, response just show HTML codes which I guess this is keycloak login page. Can you suggest what the best method to integrate login Keycloak with mobile (I use Flutterflow)?

**Comment by FlorianMeinschad** [comment]
Thank you for your answer, I tried it and can confirm the behavior.

Is this documented anywhere?
Because the rest api documentation is unfortunately a bit misleading here.

**Reply by ilseva** [reply]
hello!
did you find the correct way to implement your flow?
because I'm in the same situation, I would like to use a keycloak token to interact with directus API.

**Reply by ymys** [reply]
Me too. Still confused how to do it without webview-ing? /auth/login/keycloak response is null but there are values in raw body or headers which is keycloak login web page.

**Reply by ilseva** [reply]
At the moment, it is not possible to use the SDK with SSO: this is the result I arrived at after a few days of exploring the Directus code.
Directus is only able to validate tokens created by itself. I think using openid urls it is possible to implement token verification.
Please @aidenfoxx validate my hypothesis.

**Reply by rijkvanzanten** [reply]
The `/auth/login/:provider` endpoint — whether you use it with or without the SDK — redirects you to the login page of the provider. Once the provider redirects back to Directus a token set is generated and used throughout the rest of the Directus session from that point onwards.

**Reply by ilseva** [reply]
Right @rijkvanzanten, but this kind of implementation generates a security problem because browsers start blocking third-party cookies. 
When you use the SDK in an other application, cookies are not placed in the request headers.
I think it is time to implement true SSO, authorizing requests via the JWT issued by an external IAM.
